### ISSUE TITLE
[feature] Add build policy: missing prefer release

### ISSUE THREAD

**Issue by orgroman-cg** [conan-io/conan#8681]
Please add a new build policy or something similar:
--build=missing_prefer_release

Almost no packages if any are available for RelWithDebInfo.
For third party packages I don't need the debug information and I don't wish to compile those for RelWithDebInfo.
If this option was available this would definitely assist us with our development flow... 

In case prebuilt is available for RelWithDebInfo use the prebuilt,
But if not this policy should try taking the Release and only if both aren't available should build from source.  

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by solvingj** [comment]
Hi @orgroman-cg , I definitely can see the challenge you're trying to overcome. Unfortunately, it seems extremely unlikely to me that we would ever implement such a build policy which involves so much implicit conditional logic.  I'm not the one who would make such a decision, it's just a prediction.

With that said, there are a few ways for you to approach this problem.  None of them will be quite as convenient for you as if we added your desired business logic directly to Conan.  However, some are reasonable, and each comes with other benefits.

First, I want to make sure you are aware that you can specify any specific desired configuration item (setting/option/envvar) on a per-package basis. Many experienced Conan users still report that they were not aware that this was possible..  So, for example in a profile:

```
[settings]
*:build_type=RelWithDebInfo
openssl:build_type=Release
boost:build_type=Release
zlib:build_type=Release
```

Obviously, this approach lacks the convenience you requested. However, it is likely to be the best approach available to you, and has some stability, reproducibility, and predictability advantages you may come to appreciate later.  The power in this approach also means that the caller is capable of choosing any combination in a straightforward way, even if it's a bit tedious.  With complex things like build configurations, completeness and control can trump pure convenience at times.

There is at least one challenge with this approach which I still find "too inconvenient" for my tastes.  Several users have reported this feeling as well. That is: as the number of packages in a dependency graph grows, the more tedious it becomes to declare (and then maintain) such profiles in such a complete and explicit manner, especially when maintaining multiple profiles per package/product. Lockfiles are one way of dealing with the problem, but they are still experimental, and it's not realistic that MOST people use them at this time.  So, we definitely need to provide other options. 

I have been drafting a feature request for a few weeks for a feature which aims to improve this experience. I will link it here when I submit it. 

There's another important thing I want to make sure that you're aware of.  It is generally recommended that for any non-trivial, professional (or even semi-important) use case for Conan packages, that developers do not use binaries from conan-center. At least not beyond some POC phase.  It is recommended that Conan users take the time to build open-source recipes from Conan Center on their own infrastructure, and push them to private Conan repositories.  If you were to follow this recommendation, you could choose to build the profiles you need, such as `RelWithDebInfo` for all packages,   Note that this does not require you to setup CI.  You could perform the all the builds of the desired packages with the desired profiles manually on some local machine or build VM with just a few invocations of `conan create --build all`, and upload manually. 

In summary, please do consider (and even try) the explicit setting-per-package approach in your profile and let me know if that works for you.  Also, if you could let us know how many packages you have in your dependency graph, it would help us understand whether or not the suggested feature is relevant to you.

**Comment by orgroman-cg** [comment]
Hi @solvingj ,

First of all, thank you for the detailed and kind answer. 

Yes, I'm aware that setting per package could be specified at the profile level. 
Even so, I'm not sure profiles are the best solution for me?

How would I distribute a profile to any user that wish to build my package?
Should I make a profiles directory on my local filesystem server and let the user manually copy the profile? 
Since setting per package can't be specified at the conanfile.py I don't think that's a good solution for me. 

I already got a private jfrog artifactory server up and running,
and some of the 3rd party packages were already customized to our needs.

But if I get what you mean I should for each 3rd party create it as pkg/version@mycompany/stable
Eventually we won't use conan-center at all but for the first stage it's really time consuming. 
We already managed to get everything up and running but...
1. If a user wish to build a local package he will have to compile all the 3rd parties on RelWithDebInfo.
2. Some of those simply won't even compile correctly for RelWithDebInfo

We're trying virtual repositories to cache conan center packages but we still need to study this solution.
https://www.jfrog.com/confluence/display/JFROG/Virtual+Repositories

I think it would be a lot more user friendly to allow a quick solution to specify a global build type setting.
Thanks again for the detailed answer!

**Comment by solvingj** [comment]
>  How would I distribute a profile to any user that wish to build my package?
Should I make a profiles directory on my local filesystem server and let the user manually copy the profile?

We recommend sharing of profiles and other configuration files within a team using the feature:

`conan config install`

https://docs.conan.io/en/latest/reference/commands/consumer/config.html

You can store and maintain all your profiles for your codebase in a single GIT repository (or network file share, etc), and thus share them with a team very easily.  It can even be set to sync automatically on a schedule. 

Alternatively, if your code base consists of just a few applications,  and if you have created profiles which are highly customized for that application, you may choose to add a "profiles" folder with complete profiles dedicated to the application in the repository along with the application and recipe.  However, this strategy scale or manage well horizontally. It becomes harder to keep profiles across many such applications and repositories consistent as they change over time.  Thus, most people use `conan config install` and maintain a single repository of profiles.

**Comment by AbrilRBS** [comment]
This has now been solved in Conan 2 with the [binary `compatibility.py` plugin extension](https://docs.conan.io/2/reference/extensions/binary_compatibility.html), which lets users customize how Conan handles compatible binaries when it looks for them - right now by default it allows for different cppstd values to be compatible with eachother.

Modifying this plugin to allow for the expected behaviour described here should be straightforward - let us know if that's not the case!
